File: simulation/engine/stats.py
from collections import defaultdict

from models.modifier import Modifier, STAT_SCALE
from models.enums import Stat, Operation, Stacking


def apply_stacking(modifiers: list[Modifier]) -> list[Modifier]:
    """Collapse modifier list per stacking rules, grouped by (stat, operation)."""
    groups: dict[tuple, list[Modifier]] = defaultdict(list)
    for m in modifiers:
        groups[(m.stat, m.operation)].append(m)

    result = []
    for (stat, op), group in groups.items():
        stacking = group[-1].stacking  # use the last one's stacking rule
        if stacking == Stacking.stack:
            result.extend(group)
        elif stacking == Stacking.replace:
            result.append(group[-1])
        elif stacking == Stacking.max:
            result.append(max(group, key=lambda m: m.value))
    return result


def calculate_stat(base: int, modifiers: list[Modifier], stat: Stat = Stat.HP) -> int:
    """
    Resolve a single stat value from base + active modifiers.
    stat param: used to filter modifiers and determine floor-at-0 rule.
    Resolution: (base + flat_sum) * (100 + pct_sum) // 100, then sequential MULTIPLY.
    """
    # 1. Filter to matching stat
    stat_mods = [m for m in modifiers if m.stat == stat]

    # 2. Apply stacking rules
    stat_mods = apply_stacking(stat_mods)

    # 3. Sum FLAT_ADD and FLAT_SUB
    flat_sum = sum(m.value for m in stat_mods if m.operation == Operation.FLAT_ADD)
    flat_sum -= sum(m.value for m in stat_mods if m.operation == Operation.FLAT_SUB)

    # 4. Sum PCT_ADD and PCT_SUB
    pct_sum = sum(m.value for m in stat_mods if m.operation == Operation.PCT_ADD)
    pct_sum -= sum(m.value for m in stat_mods if m.operation == Operation.PCT_SUB)

    # 5. Apply flat+pct formula
    result = (base + flat_sum) * (100 + pct_sum) // 100

    # 6. Apply MULTIPLY modifiers sequentially
    for m in stat_mods:
        if m.operation == Operation.MULTIPLY:
            result = result * m.value // 1000

    # 7. Floor at 0 for non-HP stats (HP can be negative for death signal)
    if stat != Stat.HP:
        result = max(0, result)

    return result

File: game/src/sim/stats.ts
import { pyFloorDiv } from "./math";
import { Operation, SPEED_MIN_FLOOR, SPEED_PCT_CAP, STAT_SCALE, Stacking, Stat, type Modifier } from "./types";

export function applyStacking(modifiers: Modifier[]): Modifier[] {
  const groups = new Map<string, Modifier[]>();
  for (const modifier of modifiers) {
    const key = `${modifier.stat}\u0000${modifier.operation}`;
    const group = groups.get(key);
    if (group) {
      group.push(modifier);
    } else {
      groups.set(key, [modifier]);
    }
  }

  const result: Modifier[] = [];
  for (const group of groups.values()) {
    const stacking = group[group.length - 1]?.stacking;
    if (stacking === Stacking.stack) {
      result.push(...group);
    } else if (stacking === Stacking.replace) {
      result.push(group[group.length - 1]);
    } else if (stacking === Stacking.max) {
      result.push(group.reduce((best, current) => current.value > best.value ? current : best));
    }
  }
  return result;
}

export function calculateStat(base: number, modifiers: Modifier[], stat: Stat = Stat.HP): number {
  const statMods = applyStacking(modifiers.filter((modifier) => modifier.stat === stat));
  let flatSum = 0;
  let pctSum = 0;

  for (const modifier of statMods) {
    if (modifier.operation === Operation.FLAT_ADD) flatSum += modifier.value;
    if (modifier.operation === Operation.FLAT_SUB) flatSum -= modifier.value;
    if (modifier.operation === Operation.PCT_ADD) pctSum += modifier.value;
    if (modifier.operation === Operation.PCT_SUB) pctSum -= modifier.value;
  }

  if (stat === Stat.Speed && pctSum > SPEED_PCT_CAP) {
    pctSum = SPEED_PCT_CAP;
  }

  let result = pyFloorDiv((base + flatSum) * (100 + pctSum), 100);
  for (const modifier of statMods) {
    if (modifier.operation === Operation.MULTIPLY) {
      result = pyFloorDiv(result * modifier.value, STAT_SCALE);
    }
  }

  if (stat !== Stat.HP) {
    result = Math.max(0, result);
  }
  if (stat === Stat.Speed) {
    result = Math.max(result, SPEED_MIN_FLOOR * STAT_SCALE);
  }
  return result;
}
